realm_gpu/main.cc

```diff
@@ -0,0 +1,1152 @@
+          events.push_back(
+            RegionInstance::create_instance(war_exchange.at(graph_index).at(point), memory, bounds, field_sizes,
+                                            0 /*SOA*/, ProfilingRequestSet()));
+        }
+      }
+    }
+  }
+
+  Event::merge_events(events).wait();
+  events.clear();
+
+  Barrier sync_bar = Barrier::create_barrier(num_procs);
+
+  Barrier first_start_bar =
+      Barrier::create_barrier(num_procs, REDOP_MIN, &RedopMin::identity,
+                              sizeof(RedopMin::identity));
+  Barrier last_start_bar =
+      Barrier::create_barrier(num_procs, REDOP_MAX, &RedopMax::identity,
+                              sizeof(RedopMax::identity));
+  Barrier first_stop_bar =
+      Barrier::create_barrier(num_procs, REDOP_MIN, &RedopMin::identity,
+                              sizeof(RedopMin::identity));
+  Barrier last_stop_bar =
+      Barrier::create_barrier(num_procs, REDOP_MAX, &RedopMax::identity,
+                              sizeof(RedopMax::identity));
+
+  for (long proc_index = 0; proc_index < num_procs; ++proc_index) {
+    auto proc = procs.at(proc_index);
+
+    ShardArgs args;
+    args.proc_index = proc_index;
+    args.num_procs = num_procs;
+    args.num_fields = num_fields;
+    args.force_copies = force_copies;
+    args.gpumem = proc_gpumems[proc];
+    args.zcopymem = proc_zcopymems[proc];
+    args.sync = sync_bar;
+    args.first_start = first_start_bar;
+    args.last_start = last_start_bar;
+    args.first_stop = first_stop_bar;
+    args.last_stop = last_stop_bar;
+
+    DynamicBufferSerializer ser(4096);
+    ser << args;
+    ser << graphs;
+    ser << task_results;
+    ser << task_inputs;
+    ser << raw_exchange;
+    ser << war_exchange;
+    events.push_back(
+      proc.spawn(SHARD_TASK, ser.get_buffer(), ser.bytes_used()));
+  }
+
+  Event::merge_events(events).wait();
+  events.clear();
+
+  unsigned long long first_start;
+  {
+    first_start_bar.wait();
+#ifndef NDEBUG
+    bool ok =
+#endif
+      first_start_bar.get_result(&first_start, sizeof(first_start));
+    assert(ok);
+  }
+
+  unsigned long long last_start;
+  {
+    last_start_bar.wait();
+#ifndef NDEBUG
+    bool ok =
+#endif
+      last_start_bar.get_result(&last_start, sizeof(last_start));
+    assert(ok);
+  }
+
+  unsigned long long first_stop;
+  {
+    first_stop_bar.wait();
+#ifndef NDEBUG
+    bool ok =
+#endif
+      first_stop_bar.get_result(&first_stop, sizeof(first_stop));
+    assert(ok);
+  }
+
+  unsigned long long last_stop;
+  {
+    last_stop_bar.wait();
+#ifndef NDEBUG
+    bool ok =
+#endif
+      last_stop_bar.get_result(&last_stop, sizeof(last_stop));
+    assert(ok);
+  }
+
+  app.report_timing((last_stop - first_start)/1e9);
+}
+
+int main(int argc, char **argv)
+{
+  sleep(20);
+  Runtime rt;
+
+  rt.init(&argc, &argv);
+
+  rt.register_task(TOP_LEVEL_TASK, top_level_task);
+  rt.register_task(SHARD_TASK, shard_task);
```

Comment:
It is better to use `Processor::register_task_by_kind(Processor::XXX,...) `to register non-top-level tasks.